octodns/provider/selectel.py

```diff
@@ -0,0 +1,303 @@
+#
+#
+#
+
+from __future__ import absolute_import, division, print_function, \
+    unicode_literals
+
+from collections import defaultdict
+
+from logging import getLogger
+
+from requests import Session
+
+from ..record import Record, Update
+from .base import BaseProvider
+
+
+class SelectelAuthenticationRequired(Exception):
+    def __init__(self, msg):
+        message = 'Authorization failed. Invalid or empty token.'
+        super(SelectelAuthenticationRequired, self).__init__(message)
+
+
+class SelectelProvider(BaseProvider):
+    SUPPORTS_GEO = False
+
+    SUPPORTS = set(('A', 'AAAA', 'CNAME', 'MX', 'NS', 'TXT', 'SPF', 'SRV'))
+
+    MIN_TTL = 60
+
+    LIMIT = 50
+
+    API_URL = 'https://api.selectel.ru/domains/v1'
+
+    def __init__(self, id, token, *args, **kwargs):
+        self.log = getLogger('SelectelProvider[{}]'.format(id))
+        self.log.debug('__init__: id=%s', id)
+        super(SelectelProvider, self).__init__(id, *args, **kwargs)
+
+        self._sess = Session()
+        self._sess.headers.update({
+            'X-Token': token,
+            'Content-Type': 'application/json',
+        })
+        self._zone_records = {}
+        self._domain_list = self.domain_list()
+        self._zones = None
+
+    def _request(self, method, path, params=None, data=None):
+        self.log.debug('_request: method=%s, path=%s', method, path)
+
+        url = '{}{}'.format(self.API_URL, path)
+        resp = self._sess.request(method, url, params=params, json=data)
+
+        self.log.debug('_request: status=%s', resp.status_code)
+        if resp.status_code == 401:
+            raise SelectelAuthenticationRequired(resp.text)
+        elif resp.status_code == 404:
+            return {}
+        resp.raise_for_status()
+        if method == 'DELETE':
+            return {}
+        return resp.json()
+
+    def _get_total_count(self, path):
+        url = '{}{}'.format(self.API_URL, path)
+        resp = self._sess.request('HEAD', url)
+        return int(resp.headers['X-Total-Count'])
+
+    def _request_with_pagination(self, path, total_count):
+        result = []
+        for offset in range(0, total_count, self.LIMIT):
+            result += self._request('GET', path,
+                                    params={'limit': self.LIMIT,
+                                            'offset': offset})
+        return result
+
+    def _include_change(self, change):
+        if isinstance(change, Update):
+            existing = change.existing.data
+            new = change.new.data
+            new['ttl'] = max(self.MIN_TTL, new['ttl'])
+            if new == existing:
```

Comment:
Probably good to include a `self.log.debug` here so that if somebody runs into something that doesn't make sense to them (why is my TTL not working) they can do a `--debug` and get more info about what's happening.